src/geeknn/ordination/gnn.py

```
@@ -1,218 +1,38 @@
-import ee
+from __future__ import annotations
 
-from . import utils
-from .cca import cca
+import numpy as np
+from sknnr.transformers import CCATransformer
 
-# TODO: Maybe make dataclass if GEE can handle that
-# TODO: Don't assign self variables in train
+from ._base import Transformed
 
 
-class GNN:
+class GNN(Transformed):
+    SPP_TRANSFORM_FUNC = {
+        "SQRT": lambda x: np.sqrt(x),
+        "LOG": lambda x: np.log(x),
+        "NONE": lambda x: x,
+    }
+
     def __init__(self, k=1, spp_transform="SQRT", num_cca_axes=8, max_duplicates=None):
-        self.k = k
         self.spp_transform = spp_transform
```

Fail fast with invalid params by getting and storing the `spp_transform` func during init? 

Another option you could consider would be just taking a callable, e.g. `spp_transform=np.sqrt` - more flexible if users want to pass a custom callable, but a little less convenient if they have to import `numpy` to get `np.log`.

---

Thanks! I went with this approach.